File: swift/Sources/ActionCable/Transport.swift
import Foundation

/// What the client needs the transport to negotiate: the subprotocols its
/// protocol adapters speak, and the headers that authenticate the request — a
/// cookie or a token, since an Action Cable server authorizes the upgrade
/// request itself.
public struct DialOptions: Sendable, Hashable {
    public var subprotocols: [String]
    public var headers: [String: String]

    public init(subprotocols: [String] = [], headers: [String: String] = [:]) {
        self.subprotocols = subprotocols
        self.headers = headers
    }
}

/// Dials the network connection a client talks over. It is the seam where a
/// network handler plugs in: the built-in ``URLSessionTransport`` speaks
/// RFC 6455 through `URLSessionWebSocketTask`, and wrapping another WebSocket
/// package, or an in-memory pipe for tests, means implementing these two
/// protocols and nothing else.
public protocol Transport: Sendable {
    func dial(url: String, options: DialOptions) async throws -> any Connection
}

/// One live connection.
///
/// `read` and `write` are each called from one task at a time, but `close` may
/// be called while either is running, and must interrupt them. So must
/// cancelling the task that is reading or writing: that is how the client hangs
/// up a connection that has gone quiet.
public protocol Connection: Sendable {
    /// What the server negotiated, empty if it named none.
    var subprotocol: String { get }

    /// The next complete message. It throws once the connection is unusable,
    /// including when the reading task is cancelled.
    func read() async throws -> Data

    /// Sends one text message.
    func write(_ payload: Data) async throws

    func close() async
}

/// A connection that can say why it is hanging up. ``Connection/close()`` sends
/// a close frame with 1000 Normal Closure; `close(code:reason:)` sends one with
/// the code and reason given, for a caller with something to tell the server.
/// The built-in transport's connections implement it.
public protocol StatusClosing: Connection {
    func close(code: Int, reason: String) async
}

File: ruby/lib/actioncable_client/transport.rb
# frozen_string_literal: true

module ActionCableClient
  # Transports dial the network connection a client talks over. This is the
  # seam where a network handler plugs in: the built-in WebSocketTransport
  # speaks RFC 6455 on the standard library, and wrapping faye-websocket, a
  # WebSocket of an application's own, or an in-memory pipe for tests means
  # answering these two sets of messages and nothing else.
  #
  # A transport answers:
  #
  #   dial(url, subprotocols:, headers:) -> connection
  #
  # where subprotocols are the ones the client's protocols negotiate under and
  # headers are what authorizes the request — a cookie or a token, since an
  # Action Cable server authorizes the upgrade request itself.
  #
  # A connection answers:
  #
  #   subprotocol            what the server negotiated, nil if it named none
  #   read(timeout:)         the next complete message, as a String
  #   write(payload, timeout:)  sends one text message
  #   close                  hangs up
  #
  # #read and #write are each called from one thread at a time, but #close may
  # be called alongside either and has to interrupt it. A #read that runs out
  # of time raises TimeoutError; one on a connection that is finished raises.
  #
  # A connection that can say why it is hanging up also answers
  # close_with_status(code, reason), where #close sends 1000 Normal Closure.
  # The built-in transport's connections do.
  module Transport
  end
end
